### PR TITLE
8311610: [lworld+vector] Clean-up of vector allocation in class VectorSupport

### PR DESCRIPTION
**Pull request by XiaohongGong**
PR [1] has added the vector supports for field re-assigning after object re-allocation in `deoptimization.cpp`. The similar vector
and vector payload re-allocation code in `vectorSupport.cpp` can be removed now.

Besides, there is an issue in current vector payload allocation code in `vectorSupport.cpp`, that the multi-fields are forgotten
to be assigned after the payload object is allocated (see [2]). This method is called separately during deoptimization (see [3]).
This makes several Vector API jtreg test results incorrect. And either assigning the fields by calling `Deoptimization::reassign_fields_by_klass()` after it or directly going to the normal re-allocation path in `deoptimization.cpp` can fix the issue. I prefer the latter one. Hence the similar code in `vectorSupport.cpp` is needless and can be removed.

This patch also fixes a crash when printing out the JVM state, which is caused by the additional multi-fields passed to a safepoint. Normally, the fields number passed to a safepoint is the same with the klass's non-static fields number. But it's different for multi-fields. Only the `multifield_base` is added to the klass's non-static field list. Hence if all the un-vectorized multi-fields are passed to a safepoint, the fields number is larger than the klass's non-static fields number. And the array access at [4] throws exception due to out of range index.

[1] https://github.com/openjdk/valhalla/pull/866
[2] https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/prims/vectorSupport.cpp#L281
[3] https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/runtime/deoptimization.cpp#L1282
[4] https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/opto/callnode.cpp#L527

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8311610](https://bugs.openjdk.org/browse/JDK-8311610): [lworld+vector] Clean-up of vector allocation in class VectorSupport (**Enhancement** - P4)

### Reviewers
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/882/head:pull/882` \
`$ git checkout pull/882`

Update a local copy of the PR: \
`$ git checkout pull/882` \
`$ git pull https://git.openjdk.org/valhalla.git pull/882/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 882`

View PR using the GUI difftool: \
`$ git pr show -t 882`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/882.diff">https://git.openjdk.org/valhalla/pull/882.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/valhalla/pull/882#issuecomment-1624846490)

### CODE REVIEW

**Comment by XiaohongGong** [comment]
Hi @jatin-bhateja, could you please take a look at this PR? This is a cleanup to the vector allocation code in `vectorSupport.cpp`, which could also fix the jtreg regressions. Thanks a lot!

**Comment by jatin-bhateja** [comment]
> Hi @jatin-bhateja, could you please take a look at this PR? This is a cleanup to the vector allocation code in `vectorSupport.cpp`, which could also fix the jtreg regressions. Thanks a lot!

Hi @XiaohongGong , we can sweep out explicit handling related to vector/mask/shuffle in during deoptimization, only two additional thing which may be needed are as follows :-  
1/ Explicit larval flag setting on re-allocated vector payloads.
2/ Handling for non-flattened payloads under influence of InlineFieldMaxFlatSize.

**Comment by XiaohongGong** [comment]
> > Hi @jatin-bhateja, could you please take a look at this PR? This is a cleanup to the vector allocation code in `vectorSupport.cpp`, which could also fix the jtreg regressions. Thanks a lot!
> 
> Hi @XiaohongGong ,
> 
> We can sweep out explicit handling related to vector/mask/shuffle during deoptimization, only two additional thing which may be needed are as follows :- 1/ Explicit larval flag setting on re-allocated vector payloads. 2/ Handling for non-flattened payloads under influence of InlineFieldMaxFlatSize.
> 
> Best Regards, Jatin

Yes, I forgot the setting to larval flag. I will add this later. For 2, currently I don't know that the "final" flag influence to the flatten field. Since the payload field is final in Vector API, maybe it can always be flattened as you said in the maillist, right? Besides, I think the normal re-allocation in deoptimization.cpp should also consider the non-flattened fields as well, isn't it?

**Comment by XiaohongGong** [comment]
Besides the vector api regressions, I met lots of crashes due to nullptr of the payload field. It looks like being caused by deoptmization as well, where the payload field is not allocated I guess. With the change in this PR, the crash cases increased. Do you have any idea about this?

**Comment by jatin-bhateja** [comment]
> Besides the vector api regressions, I met lots of crashes due to nullptr of the payload field. It looks like being caused by deoptmization as well, where the payload field is not allocated I guess. With the change in this PR, the crash cases increased. Do you have any idea about this?

Can you try integrating https://github.com/openjdk/valhalla/pull/884 locally and check if it resolves the crashes.

**Comment by XiaohongGong** [comment]
> > Besides the vector api regressions, I met lots of crashes due to nullptr of the payload field. It looks like being caused by deoptmization as well, where the payload field is not allocated I guess. With the change in this PR, the crash cases increased. Do you have any idea about this?
> 
> Can you try integrating #884 locally and check if it resolves the crashes.

Thanks for the patch! I tested with it, but unfortunatelly, the crash is still there. I will look at this issue today. Thanks!

**Comment by jatin-bhateja** [comment]
> > > Besides the vector api regressions, I met lots of crashes due to nullptr of the payload field. It looks like being caused by deoptmization as well, where the payload field is not allocated I guess. With the change in this PR, the crash cases increased. Do you have any idea about this?
> > 
> > 
> > Can you try integrating #884 locally and check if it resolves the crashes.
> 
> Thanks for the patch! I tested with it, but unfortunatelly, the crash is still there. I will look at this issue today. Thanks!

> > > Besides the vector api regressions, I met lots of crashes due to nullptr of the payload field. It looks like being caused by deoptmization as well, where the payload field is not allocated I guess. With the change in this PR, the crash cases increased. Do you have any idea about this?
> > 
> > 
> > Can you try integrating #884 locally and check if it resolves the crashes.
> 
> Thanks for the patch! I tested with it, but unfortunatelly, the crash is still there. I will look at this issue today. Thanks!

Are you referring to following assertion failures which are seen without -Xshare:off
test Float256VectorTests.unsliceUnaryFloat256VectorTests(float[-i * 5]): failure
java.lang.NoClassDefFoundError: Could not initialize class jdk.internal.vm.vector.VectorSupport$VectorPayloadMF256F
	at java.base/jdk.internal.vm.vector.VectorSupport$VectorPayloadMF.newInstanceFactory(VectorSupport.java:220)
	at jdk.incubator.vector/jdk.incubator.vector.AbstractSpecies.makeDummyVectorMF(AbstractSpecies.java:369)
	at jdk.incubator.vector/jdk.incubator.vector.AbstractSpecies.dummyVectorMF(AbstractSpecies.java:327)
	at jdk.incubator.vector/jdk.incubator.vector.FloatVector$FloatSpecies.dummyVectorMF(FloatVector.java:3859)
	at jdk.incubator.vector/jdk.incubator.vector.FloatVector.fromArray(FloatVector.java:2894)
	at Float256VectorTests.unsliceUnaryFloat256VectorTests(Float256VectorTests.java:3633)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:599)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:174)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)

**Comment by XiaohongGong** [comment]
> > > > Besides the vector api regressions, I met lots of crashes due to nullptr of the payload field. It looks like being caused by deoptmization as well, where the payload field is not allocated I guess. With the change in this PR, the crash cases increased. Do you have any idea about this?
> > > 
> > > 
> > > Can you try integrating #884 locally and check if it resolves the crashes.
> > 
> > 
> > Thanks for the patch! I tested with it, but unfortunatelly, the crash is still there. I will look at this issue today. Thanks!
> 
> > > > Besides the vector api regressions, I met lots of crashes due to nullptr of the payload field. It looks like being caused by deoptmization as well, where the payload field is not allocated I guess. With the change in this PR, the crash cases increased. Do you have any idea about this?
> > > 
> > > 
> > > Can you try integrating #884 locally and check if it resolves the crashes.
> > 
> > 
> > Thanks for the patch! I tested with it, but unfortunatelly, the crash is still there. I will look at this issue today. Thanks!
> 
> Are you referring to following assertion failures which are seen without -Xshare:off
> 
> ```
> test Float256VectorTests.unsliceUnaryFloat256VectorTests(float[-i * 5]): failure
> java.lang.NoClassDefFoundError: Could not initialize class jdk.internal.vm.vector.VectorSupport$VectorPayloadMF256F
> 	at java.base/jdk.internal.vm.vector.VectorSupport$VectorPayloadMF.newInstanceFactory(VectorSupport.java:220)
> 	at jdk.incubator.vector/jdk.incubator.vector.AbstractSpecies.makeDummyVectorMF(AbstractSpecies.java:369)
> 	at jdk.incubator.vector/jdk.incubator.vector.AbstractSpecies.dummyVectorMF(AbstractSpecies.java:327)
> 	at jdk.incubator.vector/jdk.incubator.vector.FloatVector$FloatSpecies.dummyVectorMF(FloatVector.java:3859)
> 	at jdk.incubator.vector/jdk.incubator.vector.FloatVector.fromArray(FloatVector.java:2894)
> 	at Float256VectorTests.unsliceUnaryFloat256VectorTests(Float256VectorTests.java:3633)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
> 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
> 	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:599)
> 	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:174)
> 	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
> ```

I didn't see this failure without `-Xshare:off` on NEON. You could also reproduce the crash with "-XX:-EnableVectorSupport" when run a jtreg that the hardware doesn't support the vectorization for it. For example, maybe running the 512 species tests with "-XX:UseAVX=2". The crash I met looks like the null point for payload instance. Here is the log:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000ffffa8c47860, pid=263319, tid=263339
#
# JRE version: OpenJDK Runtime Environment (21.0) (fastdebug build 21-internal-git-7aa742799)
# Java VM: OpenJDK 64-Bit Server VM (fastdebug 21-internal-git-7aa742799, mixed mode, sharing, compressed oops, compressed class ptrs, g1 gc, linux-aarch64)
# Problematic frame:
# j  jdk.incubator.vector.FloatVector.uOpTemplateMF(Ljdk/incubator/vector/VectorMask;Ljdk/incubator/vector/FloatVector$FUnOp;)Ljdk/incubator/vector/FloatVector;+11 jdk.incubator.vector@21-internal
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport -p%p -s%s -c%c -d%d -P%P -u%u -g%g -- %E" (or dumping to /mnt/local/code/jtreg/jtreg-git/build/images/jtreg/jtwork/scratch/0/core.263319)
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#

```

**Comment by XiaohongGong** [comment]
Hi, the latest change fixed the offset issue for multifields during deoptimization. With this fixing, all the tests (in Int256VectorTests) can pass on NEON.  Please help to take a look. Thanks a lot!

File: src/hotspot/share/runtime/deoptimization.cpp [hunk 1]
```diff
@@ -1612,12 +1612,12 @@ int Deoptimization::reassign_fields_by_klass(InstanceKlass* klass, frame* fr, Re
       intptr_t val;
       ScopeValue* scope_field = sv->field_at(svIndex);
       StackValue* value = StackValue::create_stack_value(fr, reg_map, scope_field);
-      offset += j * type2aelembytes(type);
```
Comment (jatin-bhateja) [inline src/hotspot/share/runtime/deoptimization.cpp:1617]: Nice catch!!, It went in along with merge patch. Will run this through jtreg tests to ensure status quo or better validation status.

Best Regards,
Jatin

Comment (jatin-bhateja) [reply src/hotspot/share/runtime/deoptimization.cpp:1617]: Please also relax the condition on line # 1605 to just check for Location::vector

Comment (XiaohongGong) [reply src/hotspot/share/runtime/deoptimization.cpp:1617]: Yes, relaxing the condition is fine. And we'd better to add the assertion that `secondary_fields_count > 1` ?

Comment (XiaohongGong) [reply src/hotspot/share/runtime/deoptimization.cpp:1617]: Or, does the vector supported for 1 lanecount (e.g. 64-bit shape for `long/double` type) on x86 systems?

Comment (jatin-bhateja) [reply src/hotspot/share/runtime/deoptimization.cpp:1617]: > Or, is the vector supported for 1 lanecount (e.g. 64-bit shape for `long/double` type) on x86 systems?

Not yes, most of the vector intrinsic also fail for single lane vector operations.

Comment (XiaohongGong) [reply src/hotspot/share/runtime/deoptimization.cpp:1617]: Yeah, I also met the bad ad file crashes due to single lane for double type on x86 system. And I tried to add the assertion that `secondary_fields_count > 1`, but it crashes when running `Double64VectorTests` on x86. Anyway, I have removed this condition and assertion in the deoptimization.

**Review by jatin-bhateja — APPROVED** [review]
De-optimization cleanup looks good to me. There are still failures in LoadStore and Conversion tests which can be addressed in subsequent patches.

**Comment by XiaohongGong** [comment]
> De-optimization cleanup looks good to me. There are still failures in LoadStore and Conversion tests which can be addressed in subsequent patches.

Yes, we have to spend time looking at those issues next. Thanks for the review!

**Pull request closed without merging**